File: agents/github-action-agent/src/tonic_agent/github_checks.py
"""Optional GitHub Check runs with annotations (needs checks: write on the token)."""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from typing import Any

from .head_line_map import conflict_region_to_head_span_result
from .merge import annotated_to_conflict_file
from .models import MergeArtifact


def _api_root() -> str:
    return os.environ.get("GITHUB_API_URL", "https://api.github.com").rstrip("/")


def _request(
    method: str,
    url: str,
    token: str,
    body: dict[str, Any] | None = None,
) -> dict[str, Any] | list[Any]:
    data = None if body is None else json.dumps(body).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=data,
        headers={
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {token}",
            "X-GitHub-Api-Version": "2022-11-28",
        },
        method=method,
    )
    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            raw = resp.read().decode("utf-8")
            return json.loads(raw) if raw else {}
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"GitHub API {e.code}: {e.read().decode('utf-8', errors='replace')}") from e


def post_tonic_check(
    owner: str,
    repo: str,
    head_sha: str,
    token: str,
    artifacts: list[MergeArtifact],
    pairs: dict[str, tuple[list[str], list[str], Any]],
) -> None:
    """Create a completed check run with up to 50 annotations."""
    api = _api_root()
    annotations: list[dict[str, Any]] = []
    conflict_files = 0
    file_unmapped: dict[str, int] = {}
    file_ambiguous: dict[str, int] = {}

    for a in artifacts:
        if not a.markers_present:
            continue
        conflict_files += 1
        path = a.path
        right_lines = pairs[path][1]
        cf = annotated_to_conflict_file(path, a.annotated_lines)
        prefer_after = 0
        for reg in cf.conflicts:
            res = conflict_region_to_head_span_result(
                reg, right_lines, prefer_after_line_0=prefer_after
            )
            if res["kind"] == "unique":
                prefer_after = res["span"][1]
                h_start, h_end = res["span"]
                if len(annotations) < 50:
                    annotations.append(
                        {
                            "path": path,
                            "start_line": h_start,
                            "end_line": h_end,
                            "annotation_level": "warning",
                            "message": (
                                f"Tonic conflict ({reg.conflict_kind}) — inline review mapped"
                            ),
                        }
                    )
            elif res["kind"] == "ambiguous":
                file_ambiguous[path] = file_ambiguous.get(path, 0) + 1
            else:
                file_unmapped[path] = file_unmapped.get(path, 0) + 1

    for p in sorted(set(file_unmapped) | set(file_ambiguous)):
        if len(annotations) >= 50:
            break
        u = file_unmapped.get(p, 0)
        amb = file_ambiguous.get(p, 0)
        if u == 0 and amb == 0:
            continue
        parts: list[str] = []
        if amb:
            parts.append(f"{amb} ambiguous head span(s)")
        if u:
            parts.append(f"{u} unmapped region(s)")
        annotations.append(
            {
                "path": p,
                "start_line": 1,
                "end_line": 1,
                "annotation_level": "notice",
                "message": (
                    "Tonic: "
                    + "; ".join(parts)
                    + " — see PR issue / marker branch for full markers"
                ),
            }
        )

    conclusion = "neutral" if conflict_files else "success"
    summary = (
        f"{conflict_files} file(s) with Tonic conflict markers. "
        f"{len(annotations)} annotation(s) attached (capped at 50)."
    )
    body = {
        "name": "Tonic merge",
        "head_sha": head_sha,
        "status": "completed",
        "conclusion": conclusion,
        "output": {
            "title": "Tonic pairwise merge",
            "summary": summary,
            "annotations": annotations[:50],
        },
    }
    url = f"{api}/repos/{owner}/{repo}/check-runs"
    _request("POST", url, token, body)

File: agents/github-action-agent-node/src/githubChecks.ts
import { annotatedToConflictFile } from "@mergetonic/core";

import { conflictRegionToHeadSpanResult } from "./headLineMap";

import { githubRequestJson } from "./githubApi";

import type { MergeArtifactJson } from "./mergeReport";



type HydratePair = { leftLines: string[]; rightLines: string[] };



export async function postTonicCheck(

  owner: string,

  repo: string,

  headSha: string,

  token: string,

  artifacts: MergeArtifactJson[],

  pairs: Record<string, HydratePair>,

): Promise<void> {

  const apiRoot = (process.env.GITHUB_API_URL ?? "https://api.github.com").replace(/\/$/, "");

  const annotations: Array<{

    path: string;

    start_line: number;

    end_line: number;

    annotation_level: string;

    message: string;

  }> = [];

  let conflictFiles = 0;

  const fileUnmapped: Record<string, number> = {};

  const fileAmbiguous: Record<string, number> = {};



  for (const a of artifacts) {

    if (!a.markers_present) {

      continue;

    }

    conflictFiles += 1;

    const path = a.path;

    const rightLines = pairs[path]?.rightLines ?? [];

    const cf = annotatedToConflictFile(path, a.annotated_lines ?? []);

    let preferAfter = 0;

    for (const reg of cf.conflicts) {

      const res = conflictRegionToHeadSpanResult(reg, rightLines, preferAfter);

      if (res.kind === "unique") {

        preferAfter = res.span[1];

        const [hStart, hEnd] = res.span;

        if (annotations.length < 50) {

          annotations.push({

            path,

            start_line: hStart,

            end_line: hEnd,

            annotation_level: "warning",

            message: `Tonic conflict (${reg.conflictKind}) — inline review mapped`,

          });

        }

      } else if (res.kind === "ambiguous") {

        fileAmbiguous[path] = (fileAmbiguous[path] ?? 0) + 1;

      } else {

        fileUnmapped[path] = (fileUnmapped[path] ?? 0) + 1;

      }

    }

  }



  for (const p of Object.keys({ ...fileUnmapped, ...fileAmbiguous })) {

    if (annotations.length >= 50) {

      break;

    }

    const u = fileUnmapped[p] ?? 0;

    const amb = fileAmbiguous[p] ?? 0;

    if (u === 0 && amb === 0) {

      continue;

    }

    const parts: string[] = [];

    if (amb) {

      parts.push(`${amb} ambiguous head span(s)`);

    }

    if (u) {

      parts.push(`${u} unmapped region(s)`);

    }

    annotations.push({

      path: p,

      start_line: 1,

      end_line: 1,

      annotation_level: "notice",

      message: `Tonic: ${parts.join("; ")} — see PR issue / marker branch for full markers`,

    });

  }



  const conclusion = conflictFiles ? "neutral" : "success";

  const summary = `${conflictFiles} file(s) with Tonic conflict markers. ${annotations.length} annotation(s) (capped at 50).`;

  await githubRequestJson("POST", `${apiRoot}/repos/${owner}/${repo}/check-runs`, token, {

    name: "Tonic merge",

    head_sha: headSha,

    status: "completed",

    conclusion,

    output: {

      title: "Tonic pairwise merge",

      summary,

      annotations: annotations.slice(0, 50),

    },

  });

}

